### ISSUE TITLE
[Ldap] Mismatching configuration between LdapClient and AbstractConnection

### ISSUE THREAD

**Issue by Julien-Marcou** [symfony/symfony#19008]
# Description

When using the Ldap component of **Symfony 3.1**, binding an **LdapClient** results in an **UndefinedOptionsException**.

You can easily reproduce this bug by doing this

```php
$client = new \Symfony\Component\Ldap\LdapClient('your-ldap-host');
$client->bind(); // Throws UndefinedOptionsException
```

# What is happening

When creating an instance of LdapClient, a configuration is created using the arguments of the constructor

```php
// Symfony/Component/Ldap/LdapClient.php
// __constructor($host = null, $port = 389, $version = 3, $useSsl = false, $useStartTls = false, $optReferrals = false, LdapInterface $ldap = null)

$config = array(
    'host' => $host,
    'port' => $port,
    'version' => $version,
    'useSsl' => (bool) $useSsl,
    'useStartTls' => (bool) $useStartTls,
    'optReferrals' => (bool) $optReferrals,
);
```

and an instance of Ldap is automatically created

```php
// Symfony/Component/Ldap/LdapClient.php
// __constructor($host = null, $port = 389, $version = 3, $useSsl = false, $useStartTls = false, $optReferrals = false, LdapInterface $ldap = null)

$this->ldap = null !== $ldap ? $ldap : Ldap::create('ext_ldap', $config);
```

Which results in instanciating a new Symfony\Component\Ldap\Adapter\ExtLdap\Adapter using the configuration defined above

```php
// Symfony/Component/Ldap/Ldap.php

private static $adapterMap = array(
    'ext_ldap' => 'Symfony\Component\Ldap\Adapter\ExtLdap\Adapter',
);
```

```php
// Symfony/Component/Ldap/Ldap.php
// create($adapter, array $config = array()) method

$class = self::$adapterMap[$adapter];
return new self(new $class($config));
```

Then when calling the bind() method of the LdapClient, which call the bind() method of the Ldap, which call the getConnection() method of the Adapter, it results in instanciating a new Symfony\Component\Ldap\Adapter\ExtLdap\Connection using the configuration defined above

But this Connection defines its own configuration format

```php
// Symfony/Component/Ldap/Adapter/AbstractConnection.php
// configureOptions(OptionsResolver $resolver) method

$resolver->setAllowedTypes('host', 'string');
$resolver->setAllowedTypes('port', 'numeric');
$resolver->setAllowedTypes('connection_string', 'string');
$resolver->setAllowedTypes('version', 'numeric');
$resolver->setAllowedValues('encryption', array('none', 'ssl', 'tls'));
$resolver->setAllowedTypes('options', 'array');
```

Which does not match the original configuration defined in the LdapClient, resulting in a UndefinedOptionsException when the OptionResolver tries to resolve the given configuration.

**Comment by fabpot** [comment]
I think this has been fixed now in 3.1@dev. Can you confirm?

**Comment by csarrazi** [comment]
This should be fixed, indeed.

Also, as a reminder, the `LdapClient` class should not be used unless you need it for backward compatibility reasons.

**Comment by Julien-Marcou** [comment]
Yep, it's fixed in the dev version, I didn't see it.

LdapClient now calls the normalizeConfig() method, which maps the LdapClient constructor parameters to the AbstractConnection configuration format.

@csarrazi I didn't used LdapClient, it was for the demonstration purpose, in my case it was happening when trying to login with a "form_login_ldap" firewall.

**Comment by csarrazi** [comment]
Okay!
Thanks for the confirmation, by the way! :)

**Comment by diabl0** [comment]
@Julien-Marcou : can you share example of services.yml which uses Symfony\Component\Ldap\Ldap
Unfortutnently symfony doc and cookbooks are outdated (i'm using 3.1), and i wasn't able figure it out myself :(

**Comment by csarrazi** [comment]
The docs haven't been updated for 3.1 yet.

**Comment by csarrazi** [comment]
For an example:

```
ldap:
        class: Symfony\Component\Ldap\Ldap
        factory: [ 'Symfony\Component\Ldap\Ldap', 'create']
        arguments:
            - ext_ldap
            - host: ldap.forumsys.com
              debug: true
```

In short:

* You should use the `Ldap::create()` factory to create your service.
* There are two arguments:
    * The string associated to the adapter you wish to use. Only one is supported for now, which is `ext_ldap`.
    * The options array.

Some options are global (host, port, encryption, version, connection_string). Options specific to the implementation (`ext_ldap`) are also present in the main array (`debug`, for example, which is a shorthand for increasing verbosity for error messages), as well as the `options` sub-array (you may actually pass any option as defined in the `ConnectionOptions` class, using `lower_case`.

E.g.

```php
<?php
use Symfony\Component\Ldap\Ldap;

$ldap = Ldap::create('ext_ldap', array(
    'host' => 'my-ldap-host',
    'encryption' => 'ssl',
    'version' => 3,
    'debug' => true,
    'options' => array(
        'timelimit' => 2,
        'sizelimit' => 5,
        'network_timeout' => 10,
        'referrals' => true,
    ),
));
```

**Comment by AlimjanNL** [comment]
In Symfony 3.1.6 get LDAP instance should like this:
```
        $config=[
            "connection_string"=>"ldaps://dev-env.com",
            "port"=>555,
            "version"=>3,
            "encryption"=>'ssl',
            "debug"=>true,
        ];
$ldap = Ldap::create("ext_ldap",$config);
$ldap->bind($bindUser, $bindUserPassword);
$data=$ldap->query($ldapBaseDn,$filter);
```
config "host" key is not working.

**Comment by csarrazi** [comment]
You should not use `connection_string` in combination with either `host` or `port`.
I will update the code so that an exception is thrown if it is the case.

You should either change your configuration and use `host` instead of `connection_string` (with the value `dev-env.com`), or remove the `port` option, and only use a connection string with the port inside (i.e. `ldaps://dev-env.com:555`).

In short, `host` + `port` and `connection_string` are intended to be mutually exclusive.
